File: packages/gbnf-experiment/data/20260909T144331Z_5b46f630/ported_implementation/src/rulesBuilder/rulesBuilderTypes.ts
export abstract class InternalBase {
  abstract get type(): string;

  equals(other: unknown): boolean {
    return other instanceof InternalBase && other.constructor === this.constructor;
  }

  toString(): string {
    return `${this.type}()`;
  }
}

export abstract class InternalBaseWithValue<T> extends InternalBase {
  value: T;

  constructor(value: T) {
    super();
    this.value = value;
  }

  override equals(other: unknown): boolean {
    return (
      other instanceof InternalBaseWithValue &&
      other.constructor === this.constructor &&
      JSON.stringify(this.value) === JSON.stringify(other.value)
    );
  }

  override toString(): string {
    return `${this.type}(value=${JSON.stringify(this.value)})`;
  }
}

export abstract class InternalBaseWithInt extends InternalBaseWithValue<number> {}

export abstract class InternalBaseWithListOfInts extends InternalBaseWithValue<number[]> {
  constructor(value: number[] = []) {
    super([...value]);
  }
}

export class InternalRuleDefWithNumericValue extends InternalBaseWithInt {
  get type(): 'InternalRuleDefWithNumericValue' {
    return 'InternalRuleDefWithNumericValue';
  }
}

export class InternalRuleDefChar extends InternalBaseWithListOfInts {
  get type(): 'InternalRuleDefChar' {
    return 'InternalRuleDefChar';
  }
}

export class InternalRuleDefCharAlt extends InternalBaseWithInt {
  get type(): 'InternalRuleDefCharAlt' {
    return 'InternalRuleDefCharAlt';
  }
}

export class InternalRuleDefAlt extends InternalBase {
  get type(): 'InternalRuleDefAlt' {
    return 'InternalRuleDefAlt';
  }
}

export class InternalRuleDefCharNot extends InternalBaseWithListOfInts {
  get type(): 'InternalRuleDefCharNot' {
    return 'InternalRuleDefCharNot';
  }
}

export class InternalRuleDefCharRngUpper extends InternalBaseWithInt {
  get type(): 'InternalRuleDefCharRngUpper' {
    return 'InternalRuleDefCharRngUpper';
  }
}

export class InternalRuleDefReference extends InternalBaseWithInt {
  get type(): 'InternalRuleDefReference' {
    return 'InternalRuleDefReference';
  }
}

export class InternalRuleDefEnd extends InternalBase {
  get type(): 'InternalRuleDefEnd' {
    return 'InternalRuleDefEnd';
  }
}

export class InternalRuleDefWithoutValue extends InternalBase {
  get type(): 'InternalRuleDefWithoutValue' {
    return 'InternalRuleDefWithoutValue';
  }
}

export type InternalRuleDef =
  | InternalRuleDefChar
  | InternalRuleDefEnd
  | InternalRuleDefReference
  | InternalRuleDefCharNot
  | InternalRuleDefWithNumericValue
  | InternalRuleDefWithoutValue
  | InternalRuleDefAlt
  | InternalRuleDefCharRngUpper
  | InternalRuleDefCharAlt;

export type InternalRuleDefCharOrAltChar = InternalRuleDefChar | InternalRuleDefCharAlt;

export const isRuleDefAlt = (rule: unknown): rule is InternalRuleDefAlt =>
  rule instanceof InternalRuleDefAlt;

export const isRuleDefRef = (rule: unknown): rule is InternalRuleDefReference =>
  rule instanceof InternalRuleDefReference;

export const isRuleDefEnd = (rule: unknown): rule is InternalRuleDefEnd =>
  rule instanceof InternalRuleDefEnd;

export const isRuleDefChar = (rule: unknown): rule is InternalRuleDefChar =>
  rule instanceof InternalRuleDefChar;

export const isRuleDefCharNot = (rule: unknown): rule is InternalRuleDefCharNot =>
  rule instanceof InternalRuleDefCharNot;

export const isRuleDefCharAlt = (rule: unknown): rule is InternalRuleDefCharAlt =>
  rule instanceof InternalRuleDefCharAlt;

export const isRuleDefCharRngUpper = (
  rule: unknown,
): rule is InternalRuleDefCharRngUpper => rule instanceof InternalRuleDefCharRngUpper;

File: proposed-projects/round-trip-experiment/reverse/20260909T144331Z_5b46f630/20260910T083554Z_f6f6274b/ported_implementation/gbnf/rules_builder/rules_builder_types.py
import json
from typing import Any, List, Optional, Sequence


class InternalBase:
    @property
    def type(self) -> str:
        raise NotImplementedError

    def __eq__(self, other: object) -> bool:
        return isinstance(other, InternalBase) and type(other) is type(self)

    def __hash__(self) -> int:
        return hash(self.type)

    def __repr__(self) -> str:
        return f"{self.type}()"


class InternalBaseWithValue(InternalBase):
    def __init__(self, value: Any):
        self.value = value

    def __eq__(self, other: object) -> bool:
        return (
            isinstance(other, InternalBaseWithValue)
            and type(other) is type(self)
            and self.value == other.value
        )

    def __hash__(self) -> int:
        return hash((self.type, json.dumps(self.value)))

    def __repr__(self) -> str:
        return f"{self.type}(value={json.dumps(self.value)})"


class InternalBaseWithInt(InternalBaseWithValue):
    pass


class InternalBaseWithListOfInts(InternalBaseWithValue):
    def __init__(self, value: Optional[Sequence[int]] = None):
        super().__init__(list(value) if value is not None else [])


class InternalRuleDefWithNumericValue(InternalBaseWithInt):
    @property
    def type(self) -> str:
        return "InternalRuleDefWithNumericValue"


class InternalRuleDefChar(InternalBaseWithListOfInts):
    @property
    def type(self) -> str:
        return "InternalRuleDefChar"


class InternalRuleDefCharAlt(InternalBaseWithInt):
    @property
    def type(self) -> str:
        return "InternalRuleDefCharAlt"


class InternalRuleDefAlt(InternalBase):
    @property
    def type(self) -> str:
        return "InternalRuleDefAlt"


class InternalRuleDefCharNot(InternalBaseWithListOfInts):
    @property
    def type(self) -> str:
        return "InternalRuleDefCharNot"


class InternalRuleDefCharRngUpper(InternalBaseWithInt):
    @property
    def type(self) -> str:
        return "InternalRuleDefCharRngUpper"


class InternalRuleDefReference(InternalBaseWithInt):
    @property
    def type(self) -> str:
        return "InternalRuleDefReference"


class InternalRuleDefEnd(InternalBase):
    @property
    def type(self) -> str:
        return "InternalRuleDefEnd"


class InternalRuleDefWithoutValue(InternalBase):
    @property
    def type(self) -> str:
        return "InternalRuleDefWithoutValue"


InternalRuleDef = InternalBase


def is_rule_def_alt(rule: Any) -> bool:
    return isinstance(rule, InternalRuleDefAlt)


def is_rule_def_ref(rule: Any) -> bool:
    return isinstance(rule, InternalRuleDefReference)


def is_rule_def_end(rule: Any) -> bool:
    return isinstance(rule, InternalRuleDefEnd)


def is_rule_def_char(rule: Any) -> bool:
    return isinstance(rule, InternalRuleDefChar)


def is_rule_def_char_not(rule: Any) -> bool:
    return isinstance(rule, InternalRuleDefCharNot)


def is_rule_def_char_alt(rule: Any) -> bool:
    return isinstance(rule, InternalRuleDefCharAlt)


def is_rule_def_char_rng_upper(rule: Any) -> bool:
    return isinstance(rule, InternalRuleDefCharRngUpper)


__all__: List[str] = [
    "InternalBase",
    "InternalBaseWithInt",
    "InternalBaseWithListOfInts",
    "InternalBaseWithValue",
    "InternalRuleDef",
    "InternalRuleDefAlt",
    "InternalRuleDefChar",
    "InternalRuleDefCharAlt",
    "InternalRuleDefCharNot",
    "InternalRuleDefCharRngUpper",
    "InternalRuleDefEnd",
    "InternalRuleDefReference",
    "InternalRuleDefWithNumericValue",
    "InternalRuleDefWithoutValue",
    "is_rule_def_alt",
    "is_rule_def_char",
    "is_rule_def_char_alt",
    "is_rule_def_char_not",
    "is_rule_def_char_rng_upper",
    "is_rule_def_end",
    "is_rule_def_ref",
]
